### PR TITLE
Rebase bladedisc 20220830

### PR DESCRIPTION
**Pull request by fortianyou**
TensorFlow mainly update:
1. Extract StreamExecutors into OpenXLA
2. Extract StableHLO
3. Update new DenseArray attribute https://discourse.llvm.org/t/rfc-introduce-a-new-dense-array-attribute/63279

### CODE REVIEW

File: mhlo_builder/csrc/mlir/mhlo/builder/element_wise_binary.h [hunk 1]
```diff
@@ -28,7 +28,26 @@ static constexpr const char kCompare_GT[] = "GT";
...
+  } else if (cmp == kCompare_GE) {
+    return chlo::ComparisonDirection::GE;
+  } else if (cmp == kCompare_GT) {
+    return chlo::ComparisonDirection::GT;
+  } else if (cmp == kCompare_LE) {
+    return chlo::ComparisonDirection::LE;
+  } else if (cmp == kCompare_LT) {
+    return chlo::ComparisonDirection::LT;
+  } else {
+    assert(false && "Unhandled comparison direction.");
```
Comment (qiuxiafei) [inline mhlo_builder/csrc/mlir/mhlo/builder/element_wise_binary.h:46]: better to have a `CHECK`, `assert` will be eliminated by compiler and thus undefined value in this branch.

File: mhlo_builder/csrc/mlir/mhlo/builder/element_wise_binary.h [hunk 2]
```diff
@@ -28,7 +28,26 @@ static constexpr const char kCompare_GT[] = "GT";
 static constexpr const char kCompare_LE[] = "LE";
 static constexpr const char kCompare_LT[] = "LT";
 
-mhlo::ComparisonDirection inline getComparisonDirectionFromString(
+chlo::ComparisonDirection inline getChloComparisonDirectionFromString(
+    std::string cmp) {
```
Comment (qiuxiafei) [inline mhlo_builder/csrc/mlir/mhlo/builder/element_wise_binary.h:32]: `const std::string&` ? And why not receive `const char*`?

File: pytorch_blade/WORKSPACE [hunk 3]
```diff
@@ -126,7 +126,7 @@ new_local_repository(
 new_local_repository(
   name = "mlir-hlo",
   build_file = "third_party/torch-mlir/externals/mlir-hlo/BUILD",
-  path = "../tf_community/tensorflow/compiler/xla/mlir_hlo"
+  path = "third_party/torch-mlir/externals/mlir-hlo"
```
Comment (qiuxiafei) [inline pytorch_blade/WORKSPACE:129]: We need to ensure these two directory contains identical content, do we?
This also means we can't rebase to arbitrary commit of torch-mlir and tf_community?

Comment (fortianyou) [reply pytorch_blade/WORKSPACE:129]: Yes, they both depend on MHLO; we must keep them consistent.

File: mhlo_builder/csrc/mlir/mhlo/builder/element_wise_binary.h (hunk 1 above)
Comment (fortianyou) [reply mhlo_builder/csrc/mlir/mhlo/builder/element_wise_binary.h]: done

File: mhlo_builder/csrc/mlir/mhlo/builder/element_wise_binary.h (hunk 2 above)
Comment (fortianyou) [reply mhlo_builder/csrc/mlir/mhlo/builder/element_wise_binary.h]: done

File: pytorch_blade/WORKSPACE (hunk 3 above)
Comment (qiuxiafei) [reply pytorch_blade/WORKSPACE:129]: Every one who's going to do rebase, pls note this...

**Review by qiuxiafei — APPROVED** [review]

**Comment by fortianyou** [comment]
Mark https://github.com/alibaba/BladeDISC/pull/602

**Pull request merged**